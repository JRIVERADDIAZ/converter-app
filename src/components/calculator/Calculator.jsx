import React, { useState } from 'react'

import GenericError from '../Error/GenericError'

import { errors } from '../../helpers/errors/errors'
import { opetarionSelector } from '../../helpers/operations/operationsStorage'

const Calculator = ({ mainMetric } = props) => {

  const [metric, setMetric] = useState('select a metric')
  const [metricToConvert, setMetricToConvert] = useState('select metric to convert')
  const [metricConvertion, setConvertion] = useState()

  const [errorWindow, setErrorWindow] = useState(true)

  let err = new errors();

  const locateMetric = new opetarionSelector();

  const metricUnity = locateMetric.getMetricUnits(mainMetric);

  const onTyping = (e) => {

    const inputValue = e.target.value;

    if (isNaN(inputValue)) {
      err.addNewError(404, { message: 'Value is not a number, please try again' })
    } else {
      // If it's a valid number, remove any non-numeric characters
      const numericValue = parseFloat(inputValue.replace(/[^0-9]/g, ''));
      setConvertion(numericValue);
    }

  }

  const matcher = () => {
    if (metric === 'select a metric') {
      setMetric('Kilometer')
    }
    if (metricToConvert === 'select metric to convert') {
      setMetricToConvert('Kilometer')
    }
  }

  const calcConvertion = () => {
    matcher();

    let resultMetric = locateMetric.metricConverFormula(mainMetric, metric, metricToConvert)
    let result = resultMetric(metricConvertion)

    setConvertion(result)
    return result;

  }

  return (

    <>
      {
        !errorWindow
          ? <GenericError
            setErrorWindow={() => { setErrorWindow(!errorWindow) }}
            err={err}
          />
          : <>
            <input
              type='text'
              placeholder='0'
              pattern="^\d+$"
              value={metricConvertion}
              onInput={onTyping}
            /> <br />
            <label>

              <select
                onChange={e => setMetric(e.target.value)}
              // value={metric}
              >

                {
                  metricUnity.map((metric) => (

                    <option
                      key={metric}
                      value={metric} >
                      {metric}
                    </option>

                  ))
                }

              </select>

            </label>
            <label>

              <select
                onChange={e => setMetricToConvert(e.target.value)}
                value={metricToConvert}
              >
                {
                  metricUnity.map((metric) => (

                    <option
                      key={metric}
                      value={metric} >
                      {metric}
                    </option>

                  ))
                }
              </select><br />
            </label>

            <input
              placeholder='result converted'
              disabled={true}
              value={metricConvertion}
            /><br />

            <button onClick={calcConvertion} type="button">Convert</button>
          </>
      }

    </>
  )
}

export default Calculator