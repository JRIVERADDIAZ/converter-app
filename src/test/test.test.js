import { render } from '@testing-library/react'
import Calculator from '../components/calculator/Calculator'

test('esta prueba no deve fallar', () => {
    const { container } = render(<Calculator />)
  
    expect(container).toMatchSnapshot()

})