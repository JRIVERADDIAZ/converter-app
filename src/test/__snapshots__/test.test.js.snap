// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`esta prueba no deve fallar 1`] = `
<div>
  <input
    placeholder="unit to convert"
  />
   
  <br />
  <label>
    <select>
      <option
        value="Kilometer"
      >
        Kilometer
      </option>
      <option
        value="Hectometer"
      >
        Hectometer
      </option>
      <option
        value="Meter"
      >
        Meter
      </option>
      <option
        value="Decimeter"
      >
        Decimeter
      </option>
      <option
        value="Centimeter"
      >
        Centimeter
      </option>
      <option
        value="Milimeter"
      >
        Milimeter
      </option>
    </select>
  </label>
  <label>
    <select>
      <option
        value="Kilometer"
      >
        Kilometer
      </option>
      <option
        value="Hectometer"
      >
        Hectometer
      </option>
      <option
        value="Meter"
      >
        Meter
      </option>
      <option
        value="Decimeter"
      >
        Decimeter
      </option>
      <option
        value="Centimeter"
      >
        Centimeter
      </option>
      <option
        value="Milimeter"
      >
        Milimeter
      </option>
    </select>
    <br />
  </label>
  <input
    disabled=""
    placeholder="result converted"
  />
  <br />
  <button>
    convert
  </button>
</div>
`;
